### PR TITLE
feat: implement autoscaling

### PR DESCRIPTION
**Pull request by knave**

### CODE REVIEW

File: internal/autoscaler/workerstate.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+	if addNewPeak {
+		workload.TflopsHistogram.AddSample(float64(metrics.TflopsUsage), 1.0, metrics.Timestamp)
+		w.TflopsPeak = metrics.TflopsUsage
+	}
+
+	return true
+}
+
+func (w *WorkerState) AddVramSample(workload *WorkloadState, metrics *WorkerMetrics) bool {
```
Comment (Code2Life) [inline internal/autoscaler/workerstate.go:62]: This could be calculated by select max in Greptime flow to offload memory based calculation

File: internal/autoscaler/estimator.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+	return resourceAmountFromFloat(float64(base) * math.Pow(1.+e.multiplier/confidence, e.exponent))
+}
+
+// Returns a non-negative real number that heuristically measures how much
+// confidence the history aggregated in the WorkloadState provides.
+// For a workload producing a steady stream of samples over N days at the rate
+// of 1 sample per minute, this metric is equal to N.
+// This implementation is a very simple heuristic which looks at the total count
+// of samples and the time between the first and the last sample.
+func getConfidence(s *WorkloadState, confidenceInterval time.Duration) float64 {
```
Comment (Code2Life) [inline internal/autoscaler/estimator.go:159]: This seems not a normalized confidence factor

File: internal/autoscaler/workloadstate.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	DefaultHistogramDecayHalfLife = time.Hour * 24
+)
+
+type WorkloadState struct {
+	Namespace         string
+	Name              string
+	Resources         tfv1.Resources
+	AutoScalingConfig tfv1.AutoScalingConfig
+
+	TflopsHistogram vpa.Histogram
```
Comment (Code2Life) [inline internal/autoscaler/workloadstate.go:30]: coupled with vpa algorithm, can not extend with others like XGBoost / LightGBM / Prophet / MLP

Comment (knave) [reply internal/autoscaler/workloadstate.go]: Thanks for the review.

File: internal/autoscaler/estimator.go (hunk 2 above)
Comment (knave) [reply internal/autoscaler/estimator.go]: Thanks for the review.
